crates/fluvio/src/producer/dispatcher.rs

```diff
@@ -0,0 +1,305 @@
+                // This itself is used as the message, and ensures exactly one notification
+                drop(respond_to_producer);
+            }
+            Event::Timeout => {
+                let span = debug_span!("flush", reason = "timeout");
+                self.partition_pool.flush_all().instrument(span).await?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub(crate) struct PartitionPool {
+    /// Pool for communicating with SPUs.
+    spus: Arc<SpuPool>,
+    /// Config for producer/flushing behavior
+    config: Arc<ProducerConfig>,
+    /// A channel for sending the status of batches
+    broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+    /// Pool of individual `PartitionProducer`s, which oct independently
+    pool: HashMap<ReplicaKey, PartitionProducer>,
+}
+
+impl PartitionPool {
+    /// Create a new `PartitionPool`
+    pub fn new(
+        spus: Arc<SpuPool>,
+        config: Arc<ProducerConfig>,
+        broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+    ) -> Self {
+        Self {
+            spus,
+            config,
+            pool: Default::default(),
+            broadcast_status,
+        }
+    }
+
+    /// Buffer the given record by sending it to the appropriate `PartitionProducer`.
+    ///
+    /// If a `PartitionProducer` does not exist for this record's `ReplicaKey`, one
+    /// will be created.
+    pub async fn buffer(
+        &mut self,
+        record: AssociatedRecord,
+        respond_to_caller: Sender<ProducerMsg>,
+    ) -> Result<(), FluvioError> {
+        let partition_producer = match self.pool.get_mut(&record.replica_key) {
+            Some(partition_producer) => partition_producer,
+            None => {
+                let partition_producer = PartitionProducer::new(
+                    self.spus.clone(),
+                    self.config.clone(),
+                    record.replica_key.clone(),
+                    self.broadcast_status.clone(),
+                );
+                self.pool
+                    .insert(record.replica_key.clone(), partition_producer);
+                self.pool
+                    .get_mut(&record.replica_key)
+                    .expect("just inserted, element must exist")
+            }
+        };
+
+        partition_producer.buffer(record, respond_to_caller).await?;
+        Ok(())
+    }
+
+    /// Flush the buffers for all partitions.
+    pub async fn flush_all(&mut self) -> Result<(), FluvioError> {
+        let mut flush_futures = Vec::with_capacity(self.pool.len());
+        for (_, partition_producer) in self.pool.iter_mut() {
+            let fut = partition_producer.flush();
+            flush_futures.push(fut);
+        }
+
+        // Flush all concurrently
+        let results = futures_util::future::join_all(flush_futures).await;
+        for result in results {
+            if let Err(e) = result {
+                error!("Internal error broadcasting batch status: {}", e);
```

Comment:
so error is not propagating?

Reply:
The `PartitionProducer::flush` call will send any batch-request related errors through the `broadcast_status` channel which then ends up returning to the caller. The `Result` returned by flush itself can only ever be `Err` if there is a problem with the channel itself, which is an internal error. Our options for dealing with that kind of error are:
- Panic,
- Print the error (I'm using this option, with `error!`)
- Do nothing